### PR TITLE
Add Windows support via CMake

### PR DESCRIPTION
**Pull request by squeek502**
Allows compiling lua-gumbo on Windows via CMake (the CMakeLists.txt should also work with other platforms, but I haven't tested it).

Unsure if this is something you want, but figured I'd make the PR just in case.

Example commands to build using CMake on Windows:
```
mkdir build
cd build
cmake -G "Visual Studio 14 2015 Win64" ..
cmake --build . --config Release
```
and on Linux:
```
mkdir build && cd build
cmake ..
make
```

### CODE REVIEW

**Comment by craigbarnes** [comment]
Thanks for the PR! I'm kind of surprised that building on Windows requires so few changes to the code.

Is the intention to serve as an alternative to the makefile, rather than replacing it? The `make` build currently compiles (and tests) the parser bindings for the 5.3/5.2/5.1 ABIs in parallel (if all are installed). What does the `cmake` build do if all 3 are installed on the system? Pick a specific one?

I'm open to merging this, if you can assist with maintaining/testing it.

**Comment by squeek502** [comment]
Yes, the intention was for CMake to serve as an alternative.

CMake picks whatever Lua installation it can find (see the [find_package docs](https://cmake.org/cmake/help/v3.0/command/find_package.html), [FindLua docs](https://cmake.org/cmake/help/v3.0/module/FindLua.html) and [FindLua.cmake](https://github.com/Kitware/CMake/blob/master/Modules/FindLua.cmake)). However, it's still possible to set up testing for all Lua/LuaJIT versions via CI (e.g. [here](https://github.com/squeek502/sleep): [appveyor](https://ci.appveyor.com/project/squeek502/sleep) / [travis](https://travis-ci.org/squeek502/sleep)), or by defining the environment variables that FindLua sets before running CMake.

**Comment by craigbarnes** [comment]
That seems reasonable. The rockspec only needs a single ABI target per build too, so maybe that can be adapted to use `cmake` (for Windows support).

I'm guessing VS 2015 is the minimum version that can build the code? How complete is their C99 support? Are there any features or libc functions that need to be avoided?

**Comment by squeek502** [comment]
Unsure about the minimum MSVC version needed. C99 support is messy with MSVC as far as I can tell. It seems that since VS 2015, nearly all of C99 is supported, and before that, some portion was (see [here](https://stackoverflow.com/a/28118893) and [here](https://msdn.microsoft.com/en-us/library/hh409293.aspx#BK_CRT)).

> **C99 Conformance** Visual Studio 2015 fully implements the C99 Standard Library, with the exception of any library features that depend on compiler features not yet supported by the Visual C++ compiler (for example, <tgmath.h> is not implemented).

For Luarocks, is there a way of using different build systems per OS? Luarocks has cmake support, (`build = { type = "cmake" }`, but I'm unsure if using it also means that all platforms have to use it?

**Comment by craigbarnes** [comment]
> For Luarocks, is there a way of using different build systems per OS?

I learned a while back that you can have multiple rockspecs, each with a specific `lua` version dependency, and `luarocks` picks the right one automatically. I'm not sure if that also works with `supported_platforms`. Maybe @daurnimator knows if that would work or not...?

I don't mind using the `cmake` build in the rockspec for all platforms anyway. I use `make` for development and testing because it's what I know best, but I think it's worth switching releases and packages to use `cmake` if it means better portability.

**Comment by squeek502** [comment]
Just for reference, I think this would work for using different build system's per platform:

```
build = {
  platforms = {
    unix = {
      type = "make",
      -- current build table
    },
    windows = {
      type = "cmake",
      -- etc
    }
  }
}
```

Right now, the `CMakeLists.txt` only installs the `parse` artifact. I'll work on setting it up to support installing the entire `lua-gumbo` library so that it could be used with Luarocks.

**Comment by squeek502** [comment]
Updated the `.gitattributes` to use `eol=lf`.

File: CMakeLists.txt [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+    message( FATAL_ERROR "Variable `LUA_INCLUDE_DIR' expected to be defined.")
+  endif()
+endif()
+
+# Gumbo
+set(LIB_INCLUDE_DIR ${PROJECT_SOURCE_DIR}/lib)
+set(WINDOWS_LIB_INCLUDE_DIR ${LIB_INCLUDE_DIR}/visualc/include)
+set(LIBGUMBO_OBJ attribute char_ref error parser string_buffer string_piece svg_attrs svg_tags tag tag_lookup tokenizer utf8 util vector)
+
+add_library( attribute lib/attribute.c lib/attribute.h lib/gumbo.h lib/util.h )
```
Comment (craigbarnes) [inline CMakeLists.txt]: `add_library` seems to create a library archive file per compilation unit instead of object files. Is that a cmake thing or...?

Comment (squeek502) [reply CMakeLists.txt]: Looks like we want to use `add_library( name OBJECT files )` instead; will make that change.

Comment (squeek502) [reply CMakeLists.txt]: Scratch that, `OBJECT` types can only be included as sources for other `add_library` calls. Will look into this more and figure out a better solution.

Comment (squeek502) [reply CMakeLists.txt]: Ok, figured out how to use `OBJECT` properly. Now only object files should be created for the libgumbo parts.

**Comment by daurnimator** [comment]
> luarocks picks the right one automatically. I'm not sure if that also works with supported_platforms. Maybe @daurnimator knows if that would work or not...?

I don't think so; IIRC there are only top level manifests per lua version, not for platforms.

@hishamhm or @leafo will know more.

**Comment by craigbarnes** [comment]
@daurnimator thanks -- I tested it briefly and it didn't seem to work the same as `dependencies`, which is what I expected tbh. The `build.platforms.*` method looks like the right solution.

@squeek502 this is looking good now. The only thing I'm still unsure about is if people will use the `cmake` build (not via LuaRocks) and expect the install target to install everything. Does `FindLua.cmake` not provide directory vars for installing Lua modules and do people expect that on Windows?

**Comment by hishamhm** [comment]
> @daurnimator <https://github.com/daurnimator> thanks -- I tested it
> briefly and it didn't seem to work the same as dependencies, which is
> what I expected tbh. The build.platforms.* method looks like the right
> solution.
>
Yes, website manifests are per version, not per platform. Per-platform
manifests sound too much like a corner case for now, but we could
reconsider in the future.

Usually, `build.plaforms.*` is sufficient.

>

**Comment by squeek502** [comment]
> The only thing I'm still unsure about is if people will use the cmake build (not via LuaRocks) and expect the install target to install everything. Does FindLua.cmake not provide directory vars for installing Lua modules?

I'm unsure of the best way to handle this. One of the few examples I could find is LuaDist's cmake scripts ([example](https://github.com/LuaDist/luasystem/tree/master/cmake), but they use them for all their cmake forks). Unfortunately, their install targets don't conform with Lua's default `package.path`/`package.cpath` on any platform.

<details>
<summary>Details of LuaDist's install paths</summary>

- in `dist.cmake`: `set ( INSTALL_LIB lib CACHE PATH "Where to install libraries to." )` (note: because this gets used in the `install` commands, this becomes relative to [`CMAKE_INSTALL_PREFIX`](https://cmake.org/cmake/help/v3.0/variable/CMAKE_INSTALL_PREFIX.html))
- in `lua.cmake`: `set ( INSTALL_LMOD ${INSTALL_LIB}/lua CACHE PATH "Directory to install Lua modules." )` and `set ( INSTALL_CMOD ${INSTALL_LIB}/lua CACHE PATH "Directory to install Lua binary modules." )` which then gets used as the base for the various `install` commands

So, what this actually means:

- `INSTALL_LMOD` and `INSTALL_CMOD` are `${CMAKE_INSTALL_PREFIX}/lib/lua`
  - On unix, this resolves to `/usr/local/lib/lua`, which *almost* conforms to the default `package.path`
  - On windows, this resolves to `C:/Program Files/lib/lua`, which does not conform to the default `package.path` at all

</details>

---

So, to figure out where the files *should* be installed to, let's see what the default `package.path`/`package.cpath` actually is for each Lua version.

**Windows** (this is Lua 5.1.5, but all versions use `!` for path/cpath)
- `package.path`: `.\?.lua;!\lua\?.lua;!\lua\?\init.lua;!\?.lua;!\?\init.lua` where `!` is replaced by the directory of the Lua executable file
- `package.cpath`: `.\?.dll;!\?.dll;!\loadall.dll` where `!` is replaced by the directory of the Lua executable file

From this, we can see that it's pretty much impossible to choose a good install location for Windows, as `package.path` can change on every run of Lua. So, I think for Windows it's correct to have no real install target, or maybe have an install target relative to the source (or build?) directory. Right now, the Windows install target only copies `parse.dll` into the `gumbo` dir relative to the checked out `lua-gumbo` dir.

**Unix**
- Lua 5.1.5
  - `package.path`: `./?.lua;/usr/local/share/lua/5.1/?.lua;/usr/local/share/lua/5.1/?/init.lua;/usr/local/lib/lua/5.1/?.lua;/usr/local/lib/lua/5.1/?/init.lua`
  - `package.cpath`: `./?.so;/usr/local/lib/lua/5.1/?.so;/usr/local/lib/lua/5.1/loadall.so`
- Lua 5.2.4
  - `package.path`: `/usr/local/share/lua/5.2/?.lua;/usr/local/share/lua/5.2/?/init.lua;/usr/local/lib/lua/5.2/?.lua;/usr/local/lib/lua/5.2/?/init.lua;./?.lua`
  - `package.cpath`: `/usr/local/lib/lua/5.2/?.so;/usr/local/lib/lua/5.2/loadall.so;./?.so`
- Lua 5.3.4
  - `package.path`: `/usr/local/share/lua/5.3/?.lua;/usr/local/share/lua/5.3/?/init.lua;/usr/local/lib/lua/5.3/?.lua;/usr/local/lib/lua/5.3/?/init.lua;./?.lua;./?/init.lua`
  - `package.cpath`: `/usr/local/lib/lua/5.3/?.so;/usr/local/lib/lua/5.3/loadall.so;./?.so`

So, it looks like we want to either install everything to `/usr/local/lib/lua/<major>.<minor>` or install Lua to `/usr/local/lib/share/<major>.<minor>` and the `.so` to `/usr/local/lib/lua/<major>.<minor>`. This is not too far off the current Unix install target (`parse.so` gets installed to `/usr/local/lib/lua/<major>.<minor>/gumbo`), it just needs to be applied to the Lua files as well.

File: lib/visualc/include/strings.h [hunk 2]
```diff
@@ -0,0 +1,4 @@
+/*Dummy file to satisfy source file dependencies on Windows platform*/
+#define strcasecmp _stricmp
```
Comment (craigbarnes) [inline lib/visualc/include/strings.h]: It's ok to remove these 2 aliases. They were both replaced in commit 6dca79e5d7be5d8986fcc982b44a4fd91f533906 and won't be coming back, for the reasons mentioned in the commit.

**Comment by craigbarnes** [comment]
I hear you on the install path situation. It's a real mess, even on Unix. `pkg-config` is specifically designed for that purpose, but still only Arch and Debian package it properly.

Your last 2 commits look good anyway. So long as it builds on Windows and works via LuaRocks, it's a major improvement in portability. If you address the last point above I'll go ahead and merge.

**Comment by squeek502** [comment]
Done. :+1:

**Comment by craigbarnes** [comment]
Merged. I'll add it to the rockspec and set up some CI testing before the 0.5 release.

**Comment by craigbarnes** [comment]
@squeek502 I've removed all the calls to `add_library` for the object files. As far as I can tell, the right way to build a library is to just list all the sources in a single `add_library` call. Let me know if it's still working on Windows.

**Comment by squeek502** [comment]
Ah, good to know. Still working fine on Windows, and as a bonus it compiles much faster without the intermediary steps.

File: CMakeLists.txt [hunk 3]
```diff
@@ -0,0 +1,120 @@
...
+
+if(UNIX)
+  set(CMAKE_POSITION_INDEPENDENT_CODE ON)
+endif()
+
+# Lua
+if(WIN32)
+  find_package (Lua REQUIRED)
+else()
+  # Only the include dir matters on UNIX
```
Comment (craigbarnes) [inline CMakeLists.txt]: Doesn't this conflict with the installation rules you added in later commits?

Comment (squeek502) [reply CMakeLists.txt]: True. What really matters is that LUA_INCLUDE_DIR is valid (i.e. lua.h is found), as that implies the LUA_VERSION_* variables would have been set by find_package. It'd make more sense to check for the LUA_VERSION_ variables as well as LUA_INCLUDE_DIR than to add REQUIRED in my opinion.

Something like:
```
  if(NOT LUA_INCLUDE_DIR OR NOT LUA_VERSION_MAJOR OR NOT LUA_VERSION_MINOR)
    message( FATAL_ERROR "Path to lua.h is required, but could not be found" )
  endif()
```

**Pull request merged**